### PR TITLE
[BEAM-14306] Add unit testing to pane coder

### PR DESCRIPTION
**Pull request by jrmccluskey**
Adds unit testing to pane coder and makes tweaks to match the spec described in the runner api (https://github.com/apache/beam/blob/master/model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/beam_runner_api.proto#L922)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by jrmccluskey** [comment]
I'm planning on adding test case names to the structs and routing everything through t.Run() before merging, but that will be left for tomorrow

**Comment by jrmccluskey** [comment]
R: @riteshghorse

File: sdks/go/pkg/beam/core/graph/coder/panes.go [hunk 1]
```diff
@@ -60,11 +64,11 @@ func EncodePane(v typex.PaneInfo, w io.Writer) error {
 func NewPane(b byte) typex.PaneInfo {
 	pn := typex.NoFiringPane()
 
-	if b&0x01 == 1 {
-		pn.IsFirst = true
+	if !(b&0x02 == 2) {
+		pn.IsFirst = false
```
Comment (riteshghorse) [inline sdks/go/pkg/beam/core/graph/coder/panes.go:68]: This should be set as true?

File: sdks/go/pkg/beam/core/graph/coder/panes.go [hunk 2]
```diff
@@ -60,11 +64,11 @@ func EncodePane(v typex.PaneInfo, w io.Writer) error {
...
 	}
-	if b&0x02 == 2 {
-		pn.IsLast = true
+	if !(b&0x01 == 1) {
+		pn.IsLast = false
```
Comment (riteshghorse) [inline sdks/go/pkg/beam/core/graph/coder/panes.go:71]: This should be set as true?

Comment (riteshghorse) [inline sdks/go/pkg/beam/core/graph/coder/panes.go:71]: This should be set as true?

File: sdks/go/pkg/beam/core/graph/coder/panes_test.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+			if err != nil {
+				t.Fatalf("failed to decode pane from buffer %v, got %v", buf, err)
+			}
+			if want := input; !equalPanes(got, want) {
+				t.Errorf("got pane %v, want %v", got, want)
+			}
+		})
+	}
+}
+func TestEncodePane_bad(t *testing.T) {
```
Comment (riteshghorse) [inline sdks/go/pkg/beam/core/graph/coder/panes_test.go:142]: add empty line

Comment (riteshghorse) [inline sdks/go/pkg/beam/core/graph/coder/panes_test.go:142]:
```suggestion

func TestEncodePane_bad(t *testing.T) {
```
```suggestion
func TestEncodePane_bad(t *testing.T) {
```

Comment (riteshghorse) [inline sdks/go/pkg/beam/core/graph/coder/panes_test.go:142]:
```suggestion

func TestEncodePane_bad(t *testing.T) {
```
```suggestion
func TestEncodePane_bad(t *testing.T) {
```

Comment (riteshghorse) [reply sdks/go/pkg/beam/core/graph/coder/panes_test.go:142]: add empty line

**Review by riteshghorse — CHANGES_REQUESTED** [review]

**Comment by riteshghorse** [comment]
Small nits but everything else looks good. Thanks for catching the error in coding!

File: sdks/go/pkg/beam/core/graph/coder/panes.go (hunk 1 above)
Comment (riteshghorse) [inline sdks/go/pkg/beam/core/graph/coder/panes.go:68]: This should be set as true?

**Review by riteshghorse — APPROVED** [review]
LGTM

**Comment by riteshghorse** [comment]
R: @lostluck

**Review by lostluck — APPROVED** [review]

**Comment by lostluck** [comment]
The only thing I can think of to improve things is to include hard coded values for the encoded/decoded panes,  (eg the "no firing pane" is `0xf`) etc, so we aren't relying on the round trip for correctness.  (eg. If I say A(B(0)) = 0, all that says is A and B are inverses, but not what they do.)

**Pull request merged**